### DISCUSSION TITLE
Contravariant basis vectors in Shell Element

### DISCUSSION THREAD

**Discussion by Kavan-Khaledi** [idaholab/moose#27038] [Q&A Modules: Solid mechanics]
Dear MOOSE team,

I am trying to understand the code for the solid shell element in MOOSE. The implementation of the element follows exactly the formulations given in Dvorkin and Bathe (1984) for the MITC4 shell element. However, I ran into a problem when using the element on unstructured meshes. The element works well for structured meshes, but it gives incorrect results for unstructured meshes, and this I believe is due to incorrect contravariant basis vectors introduced to rotate the stiffness matrix. As an example, I have copied the results of simple tension and bending tests on a clamped plate with structured and unstructured meshes (the input file and the mesh file are attached). As you can see, the results of the unstructured mesh are not correct.
<img width="1216" alt="ShellElement" src="https://github.com/idaholab/moose/assets/110385277/aaa86a20-b349-48fb-9d8c-3f51c168fd32">

To solve the problem, I tested the following change in the code and now the shell element works fine for unstructured meshes. 

New results on unstructured mesh:

<img width="1188" alt="ShellElement_new" src="https://github.com/idaholab/moose/assets/110385277/2bc33a9e-7006-4889-8fab-c8edd84728f4">

**Current Contravariant basis vector calucation in `ADComputeIncrementalShellStrain`:**

`ADRankTwoTensor gmninv = local_rotation_mat.transpose() * gmninv_temp * local_rotation_mat;`
```
      (*_ge[j])[i](0, 0) = (gmninv * (*_dxyz_dxi[j])[i]) * e1;
      (*_ge[j])[i](0, 1) = (gmninv * (*_dxyz_dxi[j])[i]) * e2;
      (*_ge[j])[i](0, 2) = (gmninv * (*_dxyz_dxi[j])[i]) * e3;
      (*_ge[j])[i](1, 0) = (gmninv * (*_dxyz_deta[j])[i]) * e1;
      (*_ge[j])[i](1, 1) = (gmninv * (*_dxyz_deta[j])[i]) * e2;
      (*_ge[j])[i](1, 2) = (gmninv * (*_dxyz_deta[j])[i]) * e3;
      (*_ge[j])[i](2, 0) = (gmninv * (*_dxyz_dzeta[j])[i]) * e1;
      (*_ge[j])[i](2, 1) = (gmninv * (*_dxyz_dzeta[j])[i]) * e2;
      (*_ge[j])[i](2, 2) = (gmninv * (*_dxyz_dzeta[j])[i]) * e3;
```

**Suggested changes to calculate the contravariant basis vectors** 

```
      ADRealVectorValue gi0=((*_dxyz_deta[j])[i].cross((*_dxyz_dzeta[j])[i]))/((*_dxyz_dxi[j])[i]*((*_dxyz_deta[j])[i].cross((*_dxyz_dzeta[j])[i])));
      ADRealVectorValue gi1=((*_dxyz_dzeta[j])[i].cross((*_dxyz_dxi[j])[i]))/((*_dxyz_dxi[j])[i]*((*_dxyz_deta[j])[i].cross((*_dxyz_dzeta[j])[i])));
      ADRealVectorValue gi2=((*_dxyz_dxi[j])[i].cross((*_dxyz_deta[j])[i]))/((*_dxyz_dxi[j])[i]*((*_dxyz_deta[j])[i].cross((*_dxyz_dzeta[j])[i])));
```
```
      (*_ge[j])[i](0, 0) = gi0 * e1;
      (*_ge[j])[i](0, 1) = gi0 * e2;
      (*_ge[j])[i](0, 2) = gi0 * e3;
      (*_ge[j])[i](1, 0) = gi1 * e1;
      (*_ge[j])[i](1, 1) = gi1 * e2;
      (*_ge[j])[i](1, 2) = gi1 * e3;
      (*_ge[j])[i](2, 0) = gi2 * e1;
      (*_ge[j])[i](2, 1) = gi2 * e2;
      (*_ge[j])[i](2, 2) = gi2 * e3;
```

The above contravarient basis vectors `gi0`, `gi1`, `gi2` were calculated through:

$gi0=\dfrac{(g_s \times g_t)}{g_r.(g_s \times g_t)}$
$gi1=\dfrac{(g_t \times g_r)}{g_r.(g_s \times g_t)}$
$gi2=\dfrac{(g_r \times g_s)}{g_r.(g_s \times g_t)}$

where, $g_r, g_s, g_t$ are covariant basis vectors.

However, I have not performed any other tests. Therefore, I decided to post the issue here for discussion first.

[Shell_MITC4_test.zip](https://github.com/idaholab/moose/files/14538053/Shell_MITC4_test.zip)

Best regards,
Kavan

**Comment by GiudGiud** [comment]
@cbolisetti

**Reply by bwspenc** [reply]
Thanks for looking into this, and for your input! I'll look into this tomorrow.

Just so you know, the current version of the shell also only works for flat geometries because the rotation DOFs are in the element local coordinate system, so if there is a change of orientation from one element to another they are not compatible. I have a branch where we have a pretty major update that corrects that, but we haven't merged it in yet because we are still seeing some inconsistencies. I wonder if the changes you proposed might address those.

**Reply by Kavan-Khaledi** [reply]
@bwspenc have you been able to find time to look into this? The original MITC4 element was actually developed for non-flat shells, and I think it should also work for curved geometries with unstructured meshes if the compatibility of local coordinates between all elements is preserved.

**Reply by bwspenc** [reply]
Sorry for being so slow to get to this! I've just been swamped with other things. We have someone in our team that is going to be looking into this.

I have to admit that I don't understand how one would go about maintaining the compatibility of local coordinates between elements for unstructured meshes or curved geometries. It's been a little while since I've looked into that in depth, but the 1984 Dvorkin & Bathe paper didn't really give any information on that as I recall. Maybe I'm missing something, but on a curved geometry, it seems like the rotational coordinates, which I understand need to be in the local plane of the element, would inherently be different for two neighboring elements if their orientations are different. That would be a problem for any nodes that are connected to multiple elements oriented on different planes.

We have a set of changes that have been in the works for a while that modify the element to have 3 rotational DOFs per node, aligned with the Cartesian axes, and then transforms those to a local coordinate system unique to each element. That does result in a zero-energy drilling DOF if you are modeling a flat surface, but we found that can be suppressed easily by giving it a penalty rotational stiffness. They're on this branch:
https://github.com/bwspenc/moose/tree/shell_new

I'm interested in your thoughts on this.

**Comment by Kavan-Khaledi** [comment]
@bwspenc First of all, thanks for sharing the new shell formulation. 
FYI, following the discussion #23874, we decided to extend the existing four-noded shell element in MOOSE to a six-noded triangular element, primarily for applications in modeling support systems in tunnel construction. Another requirement for our application was adding orthotropic elasticity to the shell model. These two requirements, i.e. triangular geometry and anisotropic material, necessitated a slight deviation from Dvorkin’s paper by introducing a compatible material coordinate system. This not only maintains the compatibility of local stresses but also makes the back-calculation of local force and moment resultants possible. To clarify, let me summarize the coordinate systems in Dvorkin’s paper and our modifications. This might be relevant to your new shell formulation as well. Please correct me if I have misunderstood anything.
The following coordinate systems are used for the MITC4 shell element (I am trying to use the same terminology as used in the paper):

- **The stationary Cartesian coordinate system** $x_1−x_2−x_3$: This is the global coordinate system in which the nodal locations and displacements are described (equations 1 and 2).

- **The local nodal coordinate system** $V_1−V_2−V_n$ : Three perpendicular unit vectors are defined at each node. The director vector $V_n$ is the node normal vector. The other two unit vectors, $V_1$ and $V_2$, are the axes around which the rotational degrees of freedom are defined at each node. In Dvorkin’s paper, a simple rule is used to maintain the compatibility of in-plane vectors $V_1$ and $V_2$ between elements. This convention is also applicable to curved geometries and unstructured meshes. The convention is described in the caption of Figure 1 and has been correctly implemented in MOOSE.

![image](https://github.com/idaholab/moose/assets/110385277/5ae94f36-56a1-45c2-a666-38a65293d8e2)

- **Covariant base vectors** $g_1,g_2,g_3 (g_i=dx/dr_i)$: These base vectors are used to define the Green-Lagrange strain components in the convected coordinates ($r_1,r_2,r_3$), as shown in equation 20. These strains are not compatible between elements because the orientation of $g_1,g_2,g_3$ varies from one element to another in curved geometries (or unstructured meshes). However, they become compatible once transformed to the stationary global coordinate system through the element Jacobian transformation (as implemented in MOOSE).

`covariant_transformation_matrix[j])[i] = J`

The transformation of stress to global coordinate is through the inverse of Jacobean.

`(*_contravariant_transformation_matrix[j])[i] = (*_covariant_transformation_matrix[j])[i].inverse();`

- The tricky part in the paper is hidden in the definition of the fourth-order contravariant stiffness matrix, where the elasticity tensor is transformed to a local coordinate system with orthonormal base vectors $\hat{e}_1,\hat{e}_2,\hat{e}_3$, i.e.

![image](https://github.com/idaholab/moose/assets/110385277/1c77ba70-fbaa-4bc6-9a99-15ce0c3d204f)

The $g^1, g^2, g^3$ are contravariant base vectors of the convected coordinates $r_1,r_2,r_3$.
 I must admit, the mathematics of this part is beyond my knowledge, especially equations 11 and 12 in the paper. However, what I understand is that by multiplying this elasticity tensor with the convected strain tensor, the stress components are obtained in the $\hat{e_{i}}$ coordinate system. The contravariant vectors $g^1, g^2, g^3$ that I showed in my first post above have an important property, $g^i . g_j = \delta_{ij}$. This property brings the local stresses to the $\hat{e}_1,\hat{e}_2,\hat{e}_3$ coordinate.

Dvorkin’s paper suggests the following method for calculating the $\hat{e}_1,\hat{e}_2,\hat{e}_3$ unit vectors:

![image](https://github.com/idaholab/moose/assets/110385277/388c8da7-a315-43a7-a3bb-c8b27891b983)

However, If we use the above approach for curved geometries, we lose the compatibility of $\hat{e}_i$ vectors between elements because the $g_i$ base vectors are incompatible. This is also a problem when using anisotropic shells. Therefore, we changed this and incorporated two options in our triangular shell element:

_Option 1_: $\hat{e}_3$ is the normal vector to the shell element. The user defines a vector in the input file (in the strain block). This user-defined vector is projected onto the shell plane to construct $\hat{e}_1$. Then, $\hat{e}_2=\hat{e}_3\times\hat{e}_1$

_Option 2_: If the user does not provide any vector, the same method as described above in Figure 1 is used to preserve the compatibility of local coordinates.
Now, having  compatible base vectors $\hat{e}_1,\hat{e}_2,\hat{e}_3$  at the element level, we can transform the global stresses to this local coordinate system and back-calculate the force and moment resultants.

Sorry for the long post!!
